File: model/src/main/scala/ru/mipt/acsl/decode/model/naming/Namespace.scala
package ru.mipt.acsl.decode.model.naming

import ru.mipt.acsl.decode.model.{HasInfo, LocalizedString, Referenceable}
import ru.mipt.acsl.decode.model.component.Component
import ru.mipt.acsl.decode.model.registry.Measure
import ru.mipt.acsl.decode.model.types.Alias
import ru.mipt.acsl.decode.model.types.{Const, DecodeType}

import scala.collection.mutable

/**
  * @author Artem Shein
  */

/**
  * Mutable Namespace
  */
trait Namespace extends Referenceable with Container with HasName with HasInfo {

  def alias: Alias.NsNs

  /**
    * Get measures of the current namespace
    *
    * @return Seq of [[Measure]]
    */
  def measures: Seq[Measure] = objects.filter(_.isInstanceOf[Measure]).map(_.asInstanceOf[Measure])

  /**
    * Get types of current namespace
    *
    * @return Seq of [[DecodeType]]
    */
  def types: Seq[DecodeType] = objects.filter(_.isInstanceOf[DecodeType]).map(_.asInstanceOf[DecodeType])

  /**
    * Get subset of current namespaces
    *
    * @return Seq of [[Namespace]]
    */
  def subNamespaces: Seq[Namespace] = objects.flatMap{
    case n: Namespace => Some(n)
    case _ => None
  }

  /**
    * Get parent [[Namespace]] of current namespace
    *
    * @return Parent [[Namespace]] if it defined, otherwise - None
    */
  def parent: Option[Namespace]

  def parent_=(ns: Option[Namespace]): Unit

  def isRoot: Boolean = name == Namespace.RootName

  override def name: ElementName = alias.name

  override def info: LocalizedString = alias.info

  /**
    * Get components of current namespace
    *
    * @return Seq of [[Component]]
    */
  def components: Seq[Component] = objects.flatMap {
    case c: Component => Seq(c)
    case _ => Seq.empty
  }

  def alias(name: ElementName): Option[Alias] = objects.flatMap {
    case a: Alias if a.name == name => Seq(a)
    case _ => Seq.empty
  } match {
    case s if s.size == 1 => Some(s.head)
    case _ => None
  }

  def aliases: Seq[Alias] = objects.flatMap {
    case a: Alias => Seq(a)
    case _ => Seq.empty
  }

  def consts: Seq[Const] = objects.filter(_.isInstanceOf[Const]).map(_.asInstanceOf[Const])

  def rootNamespace: Namespace = parent.map(_.rootNamespace).getOrElse(this)

  /**
    * Get all components of current namespace
    *
    * @return Seq of [[Component]]
    */
  def allComponents: Seq[Component] = components ++ subNamespaces.flatMap(_.allComponents)

  /**
    * Get all namespaces
    *
    * @return Seq of [[Namespace]]
    */
  def allNamespaces: Seq[Namespace] = this +: subNamespaces.flatMap(_.allNamespaces)

  /**
    * Get fully qualified name
    *
    * @return [[Fqn]]
    */
  def fqn: Fqn = {
    val parts = mutable.Buffer[ElementName]()
    var currentNamespace = this
    while (currentNamespace.parent.isDefined) {
      if (!currentNamespace.isRoot) {
        parts += currentNamespace.name
      }
      currentNamespace = currentNamespace.parent.get
    }
    if (!currentNamespace.isRoot) {
      parts += currentNamespace.name
    }
    Fqn(parts.reverse)
  }

}

object Namespace {

  val RootName = ElementName.newFromMangledName("%root")

  private class NamespaceImpl(var alias: Alias.NsNs, var parent: Option[Namespace],
                     var objects: Seq[Referenceable])
    extends Namespace

  def apply(alias: Alias.NsNs, parent: Option[Namespace] = None, objects: Seq[Referenceable] = Seq.empty): Namespace =
    new NamespaceImpl(alias, parent, objects)

  def newRoot: Namespace = {
    val alias = Alias.NsNs(RootName, LocalizedString.empty)(null, null)
    alias.obj = Namespace(alias, None)
    alias.parent = alias.obj
    alias.obj
  }

  // TODO: refactoring -- use fold instead
  def apply(fqn: Fqn, rootNamespace: Namespace, info: LocalizedString): Namespace = {
    var currentNamespace = rootNamespace
    val size = fqn.size
    for ((part, i) <- fqn.parts.zipWithIndex) {
      val alias = Alias.NsNs(part, if (i == size - 1) info else LocalizedString.empty)(currentNamespace, null)
      val ns = Namespace(alias, Some(currentNamespace))
      alias.obj = ns
      currentNamespace.objects ++= Seq(ns, alias)
      currentNamespace = ns
    }
    currentNamespace
  }

}

File: model/src/main/java/ru/mipt/acsl/decode/model/naming/Namespace.java
package ru.mipt.acsl.decode.model.naming;

import com.google.common.collect.Lists;
import org.jetbrains.annotations.Nullable;
import ru.mipt.acsl.decode.model.*;
import ru.mipt.acsl.decode.model.component.Component;
import ru.mipt.acsl.decode.model.proxy.MaybeProxyNamespace;
import ru.mipt.acsl.decode.model.registry.Language;
import ru.mipt.acsl.decode.model.registry.Measure;
import ru.mipt.acsl.decode.model.types.Alias;
import ru.mipt.acsl.decode.model.types.Const;
import ru.mipt.acsl.decode.model.types.DecodeType;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * Created by metadeus on 06.06.16.
 */
public interface Namespace extends Container, HasName, HasInfo, HasAlias {

    ElementName ROOT_NAME = ElementName.newInstanceFromMangledName("%root");

    static Namespace newInstance(Alias.NsNs alias, List<Referenceable> objects) {
        return new NamespaceImpl(alias, objects);
    }

    static Namespace newInstanceRoot() {
        Alias.NsNs alias = new Alias.NsNs(ROOT_NAME, Collections.emptyMap(), null, null);
        alias.obj(MaybeProxyNamespace.newInstance(Namespace.newInstance(alias, Lists.newArrayList(alias))));
        alias.parent(alias.obj().obj());
        return alias.obj().obj();
    }

    static Namespace newInstance(Fqn fqn, Namespace rootNamespace, Map<Language, String> info) {
        Namespace currentNamespace = rootNamespace;
        int size = fqn.size();
        int index = 0;
        for (ElementName part : fqn.getParts()) {
            Alias.NsNs alias = new Alias.NsNs(part, index == size - 1 ? info : Collections.emptyMap(), currentNamespace, null);
            Namespace ns = Namespace.newInstance(alias, new ArrayList<>());
            alias.obj(MaybeProxyNamespace.newInstance(ns));
            List<Referenceable> objects = currentNamespace.objects();
            objects.add(ns);
            objects.add(alias);
            currentNamespace = ns;
        }
        return currentNamespace;
    }

    @Override
    Alias.NsNs alias();

    /**
     * Get measures of the current namespace
     *
     * @return List of [[Measure]]
     */
    default List<Measure> measures() {
        return filterByClass(Measure.class);
    }

    /**
     * Get types of current namespace
     *
     * @return List of [[DecodeType]]
     */
    default List<DecodeType> types() {
        return filterByClass(DecodeType.class);
    }

    /**
     * Get subset of current namespaces
     *
     * @return Seq of [[Namespace]]
     */
    default List<Namespace> subNamespaces() {
        return filterByClass(Namespace.class);
    }

    /**
     * Get parent [[Namespace]] of current namespace
     *
     * @return Parent [[Namespace]] if it defined, otherwise - None
     */
    Optional<Namespace> parent();

    void setParent(@Nullable Namespace ns);

    default boolean isRoot() {
        return name().equals(ROOT_NAME);
    }

    default ElementName name() {
        return alias().name();
    }

    default Map<Language, String> getInfo() {
        return alias().info();
    }

    /**
     * Get components of current namespace
     *
     * @return Seq of [[Component]]
     */
    default List<Component> components() {
        return filterByClass(Component.class);
    }

    default Optional<Alias> alias(ElementName name) {
        List<Alias> aliases = objects().stream()
                .flatMap(a -> a instanceof Alias && ((Alias)a).name().equals(name)
                        ? Stream.of((Alias) a)
                        : Stream.empty())
                .collect(Collectors.toList());
        return aliases.size() == 1 ? Optional.of(aliases.get(0)) : Optional.empty();
    }

    default List<Alias> aliases() {
        return filterByClass(Alias.class);
    }

    default List<Const> consts() {
        return filterByClass(Const.class);
    }

    default Namespace rootNamespace() {
        return parent().map(Namespace::rootNamespace).orElse(this);
    }

    /**
     * Get all components of current namespace
     *
     * @return Seq of [[Component]]
     */
    default List<Component> allComponents() {
        List<Component> result = subNamespaces().stream().flatMap(n -> n.allComponents().stream())
                .collect(Collectors.toList());
        result.addAll(components());
        return result;
    }

    /**
     * Get all namespaces
     *
     * @return Seq of [[Namespace]]
     */
    default List<Namespace> allNamespaces() {
        List<Namespace> result = subNamespaces().stream().flatMap(n -> n.allNamespaces().stream())
                .collect(Collectors.toList());
        result.add(this);
        return result;
    }

    /**
     * Get fully qualified name
     *
     * @return [[Fqn]]
     */
    default Fqn fqn() {
        ArrayList<ElementName> parts = new ArrayList<>();
        Namespace currentNamespace = this;
        while (currentNamespace.parent().isPresent()) {
            Namespace parent = currentNamespace.parent().get();
            if (!currentNamespace.isRoot()) {
                parts.add(currentNamespace.name());
            }
            currentNamespace = parent;
        }
        if (!currentNamespace.isRoot()) {
            parts.add(currentNamespace.name());
        }
        Collections.reverse(parts);
        return Fqn.newInstance(parts);
    }

    @Override
    default <T> T accept(ContainerVisitor<T> visitor) {
        return visitor.visit(this);
    }

}
